credentials/alts/internal/handshaker/service/service_test.go

```diff
@@ -30,14 +30,12 @@ const (
 )
 
 func TestDial(t *testing.T) {
-	defer func() func() {
-		temp := hsDialer
-		hsDialer = func(target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-			return &grpc.ClientConn{}, nil
-		}
-		return func() {
-			hsDialer = temp
-		}
+	temp := hsDialer
+	hsDialer = func(target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+		return &grpc.ClientConn{}, nil
+	}
+	defer func() {
+		hsDialer = temp
```

Comment:
I believe this whole override is unnecessary for this test. Thought?

Reply:
For this test, we are just mocking the Dial method [here](https://github.com/grpc/grpc-go/pull/7301/files#diff-ab1d1dab791155cb2b047170ae2424da21a7b08c15d9a30bb7f7d795e3e00585R41) otherwise it will use the orig. grpc.Dial method and since the testAddress1 and testAddress2 declared in the test are just random, it will throw error. At the end, we are just reverting back the value of hsDialer, although its not needed as it won't be used later but better reverting the value back. WDYT?

Reply:
mmm Where would it error in the test?

Reply:
Oh yeah you're right...I was refering to [this](https://github.com/grpc/grpc-go/blob/master/credentials/alts/internal/handshaker/service/service.go#L38) earlier. Makes sense, thanks for pointing out. Updated the code.

Reply:
@arvindbr8 i still think this might be an issue. If you see [here](https://github.com/grpc/grpc-go/blob/master/credentials/alts/internal/handshaker/service/service.go#L38), we are setting the value of Dial method we are using in TestDial. So since we have the dummy endpoint, it will throw error. I have tried this locally.
I have reverted back that lines of code which mocks the Dial method.

Reply:
There is no need to do any mocking in this test whatsoever. Could you consider making these changes to simplify the test even more.

```
easwars@minetop: grpc-go$ git diff --cached
diff --git a/credentials/alts/internal/handshaker/service/service.go b/credentials/alts/internal/handshaker/service/service.go
index e1cdafb9..b3af0359 100644
--- a/credentials/alts/internal/handshaker/service/service.go
+++ b/credentials/alts/internal/handshaker/service/service.go
@@ -34,8 +34,6 @@ var (
        // to a corresponding connection to a hypervisor handshaker service
        // instance.
        hsConnMap = make(map[string]*grpc.ClientConn)
-       // hsDialer will be reassigned in tests.
-       hsDialer = grpc.Dial
 )
 
 // Dial dials the handshake service in the hypervisor. If a connection has
@@ -50,7 +48,7 @@ func Dial(hsAddress string) (*grpc.ClientConn, error) {
                // Create a new connection to the handshaker service. Note that
                // this connection stays open until the application is closed.
                var err error
-               hsConn, err = hsDialer(hsAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+               hsConn, err = grpc.Dial(hsAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
                if err != nil {
                        return nil, err
                }
diff --git a/credentials/alts/internal/handshaker/service/service_test.go b/credentials/alts/internal/handshaker/service/service_test.go
index a2c6a407..3e76b1f4 100644
--- a/credentials/alts/internal/handshaker/service/service_test.go
+++ b/credentials/alts/internal/handshaker/service/service_test.go
@@ -21,7 +21,6 @@ package service
 import (
        "testing"
 
-       "google.golang.org/grpc"
        "google.golang.org/grpc/internal/grpctest"
 )
 
@@ -42,22 +41,13 @@ const (
 // If a connection has already been established, this function returns it.
 // Otherwise, a new connection is created.
 func (s) TestDial(t *testing.T) {
-       temp := hsDialer
-       hsDialer = func(target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-               return &grpc.ClientConn{}, nil
-       }
-       defer func() {
-               hsDialer = temp
-       }()
        // First call to Dial, it should create a connection to the server running
        // at the given address.
        conn1, err := Dial(testAddress1)
        if err != nil {
                t.Fatalf("first call to Dial(%v) failed: %v", testAddress1, err)
        }
-       if conn1 == nil {
-               t.Fatalf("first call to Dial(%v)=(nil, _), want not nil", testAddress1)
-       }
+       defer conn1.Close()
        if got, want := hsConnMap[testAddress1], conn1; got != want {
                t.Fatalf("hsConnMap[%v]=%v, want %v", testAddress1, got, want)
        }
@@ -79,9 +69,7 @@ func (s) TestDial(t *testing.T) {
        if err != nil {
                t.Fatalf("third call to Dial(%v) failed: %v", testAddress2, err)
        }
-       if conn3 == nil {
-               t.Fatalf("third call to Dial(%v)=(nil, _), want not nil", testAddress2)
-       }
+       defer conn3.Close()
        if got, want := hsConnMap[testAddress2], conn3; got != want {
                t.Fatalf("hsConnMap[%v]=%v, want %v", testAddress2, got, want)
        }
```

Reply:
Updated the PR